### PR TITLE
76 terrain generation

### PR DESCRIPTION
**Pull request by hogsy**
Introduces fundamental changes to ensure that new maps can be created and height-map data can be interpreted in such a way as to generate the terrain.

At the moment the red channel is dedicated to height and green is for tile index.

Importing a terrain image will require a path to be typed in for the image you want to load.

Textures are now all loaded in by incrementing through 0 to 256 under the `<mapname>/tiles/` directory. Being done this way purely for the sake of allowing for editor functionality in the future (e.g. drawing tile textures etc.)

Might need to consider providing a `tilePath` variable to manifest so maps can share tile sets. Needs more thought...

Related ticket is #76

### CODE REVIEW

File: src/engine/game/actors/actor_model.cpp [hunk 1]
```diff
@@ -23,21 +23,19 @@ AModel::AModel() : Actor() {}
...
-
-    plDrawModel(model_);
-  }
+  PLMatrix4 mrot = plRotateMatrix4(angles_.x, PLVector3(1, 0, 0));
+  mrot = plMultiplyMatrix4(mrot, plRotateMatrix4(angles_.y, PLVector3(0, 1, 0)));
+  mrot = plMultiplyMatrix4(mrot, plRotateMatrix4(angles_.z, PLVector3(0, 0, 1)));
+  model_->model_matrix = plMultiplyMatrix4(mrot, plTranslateMatrix4(position_));
+  plDrawModel(model_);
 
   Actor::Draw();
```
Comment (hogsy) [inline src/engine/game/actors/actor_model.cpp]: Needs to be moved to the top before the if check, as base class might implement debugging functionality.

File: src/engine/graphics/texture_atlas.cpp [hunk 2]
```diff
@@ -45,14 +45,18 @@ void TextureAtlas::AddImage(const std::string &path, bool absolute) {
   } else {
     snprintf(full_path, sizeof(full_path) - 1, "%s", u_find2(path.c_str(), supported_image_formats, false));
   }
+
   auto* img = static_cast<PLImage *>(u_alloc(1, sizeof(PLImage), true));
   if(!plLoadImage(full_path, img)) {
-    Error("Failed to load image (%s)!\n", plGetError());
+    u_free(img);
+    LogWarn("Failed to load image (%s)!\n", plGetError());
```
Comment (hogsy) [inline src/engine/graphics/texture_atlas.cpp]: Don't throw the error message here, let the caller deal with that when it returns false. This way we don't get a warning message every time a map gets loaded.

File: src/engine/terrain.cpp [hunk 3]
```diff
@@ -44,13 +45,19 @@ const static unsigned int chunk_indices[96] = {
...
-}
+Terrain::Terrain(const std::string& tileset) {
+  // attempt to load in the atlas sheet
+  atlas_ = new TextureAtlas(512, 8);
+  for(unsigned int i = 0; i < 256; ++i) {
+    if(!atlas_->AddImage(tileset + std::to_string(i))) {
+      break;
+    }
+  }
+  atlas_->Finalize();
```
Comment (hogsy) [inline src/engine/terrain.cpp]: Finalize function should throw a warning if there aren't any images loaded. Need to double check this!
In such a case, texture pointer needs to be our default texture.

File: src/engine/terrain.cpp [hunk 4]
```diff
@@ -126,7 +133,10 @@ void Terrain::GenerateModel(Chunk* chunk, const PLVector2& offset) {
       const Tile* current_tile = &chunk->tiles[tile_x + tile_y * TERRAIN_CHUNK_ROW_TILES];
 
       float tx_x, tx_y, tx_w, tx_h;
-      atlas_->GetTextureCoords(current_tile->texture, &tx_x, &tx_y, &tx_w, &tx_h);
+      if(atlas_ == nullptr || !atlas_->GetTextureCoords(std::to_string(current_tile->texture), &tx_x, &tx_y, &tx_w, &tx_h)) {
```
Comment (hogsy) [inline src/engine/terrain.cpp]: The atlas is always created, so it should never be null. This needs removing.

File: src/engine/terrain.cpp [hunk 5]
```diff
@@ -174,7 +184,11 @@ void Terrain::GenerateModel(Chunk* chunk, const PLVector2& offset) {
     }
   }
 
-  chunk_mesh->texture = atlas_->GetTexture();
+  //if(atlas_ == nullptr) {
```
Comment (hogsy) [inline src/engine/terrain.cpp]: Ditto to the above, just needs to be `chunk_mesh->texture = atlas_->GetTexture()` again.

File: src/engine/terrain.h [hunk 6]
```diff
@@ -63,7 +62,7 @@ class Terrain {
 
     unsigned int slip{0}; // e.g. full, bottom or left?
 
-    std::string texture;
+    uint32_t texture{0};
```
Comment (hogsy) [inline src/engine/terrain.h]: Should be an unsigned char.

File: src/engine/terrain.cpp [hunk 7]
```diff
@@ -363,3 +289,167 @@ void Terrain::Draw() {
...
+  }
+
+  // Each channel is encoded with specific data
+  // red = height
+  // green = texture
+
+  unsigned int chan_length = image.width * image.height;
+
+  auto* rchan = static_cast<float*>(u_alloc(chan_length, sizeof(float), true));
+  memset(rchan, 0, chan_length);
```
Comment (solemnwarning) [inline src/engine/terrain.cpp]: Unnecessary memset?

File: src/engine/terrain.cpp [hunk 8]
```diff
@@ -363,3 +289,167 @@ void Terrain::Draw() {
...
+  uint8_t* pixel = image.data[0];
+  for (unsigned int i = 0; i < chan_length; ++i) {
+    rchan[i] = static_cast<int>(*pixel) * multiplier; //(static_cast<int>(*pixel) - 127) * 256;
+    pixel += 4;
+  }
+
+  auto* gchan = static_cast<uint8_t*>(u_alloc(chan_length, sizeof(uint8_t), true));
+  memset(gchan, 0, chan_length);
```
Comment (solemnwarning) [inline src/engine/terrain.cpp]: Ditto

File: src/engine/terrain.cpp [hunk 9]
```diff
@@ -363,3 +289,167 @@ void Terrain::Draw() {
...
+          current_tile->shading[1] =
+          current_tile->shading[2] =
+          current_tile->shading[3] = 255;
+        }
+      }
+
+      // Find the maximum and minimum points
+      for (auto& tile : current_chunk.tiles) {
+        for (float i : tile.height) {
+          if (i > max_height_) {
```
Comment (solemnwarning) [inline src/engine/terrain.cpp]: `max_height_` should be initialised to either its minimum possible value or one of the tiles before this loop I think.

File: src/engine/terrain.cpp [hunk 10]
```diff
@@ -363,3 +289,167 @@ void Terrain::Draw() {
...
+            max_height_ = i;
+          }
+          if (i < min_height_) {
+            min_height_ = i;
```
Comment (solemnwarning) [inline src/engine/terrain.cpp]: Same for `min_height_`

File: src/engine/terrain.cpp (hunk 7 above)
Comment (hogsy) [reply src/engine/terrain.cpp]: fixed

File: src/engine/terrain.cpp (hunk 8 above)
Comment (hogsy) [reply src/engine/terrain.cpp]: also fixed

File: src/engine/terrain.cpp (hunk 9 above)
Comment (hogsy) [reply src/engine/terrain.cpp]: fixed

File: src/engine/terrain.cpp (hunk 10 above)
Comment (hogsy) [reply src/engine/terrain.cpp]: fixed

**Pull request merged**